test/e2e/middleware-can-set-the-matcher-in-its-config/index.test.ts

```diff
@@ -0,0 +1,97 @@
+import { createNext } from 'e2e-utils'
+import { NextInstance } from 'test/lib/next-modes/base'
+import { fetchViaHTTP } from 'next-test-utils'
+
+describe('Middleware can set the matcher in its config', () => {
+  describe('using a path pattern', () => {
+    let next: NextInstance
+
+    beforeAll(async () => {
+      next = await createNext({
+        files: {
+          'pages/index.js': `
+          export default function Page() { 
+            return <p>root page</p>
+          } 
+        `,
+          'pages/with-middleware/index.js': `
+          export default function Page() { 
+            return <p>This should run the middleware</p>
+          } 
+        `,
+          'middleware.js': `
+          import { NextResponse } from 'next/server'
+          export const config = {
+            matcher: '/with-middleware/:path*'
```

Comment:
I wonder if this should be `matching`, because we might want to also allow:

```ts
export const config = {
  matching: ["/with-middleware/:path*", /middleware/]
}
```